ros_speech_recognition/scripts/speech_recognition_node.py

```diff
@@ -286,12 +287,16 @@ def audio_cb(self, _, audio):
         try:
             rospy.logdebug("Waiting for result... (Sent %d bytes)" % len(audio.get_raw_data()))
             result = self.recognize(audio)
+            confidence = 1.0
+            if self.engine == Config.SpeechRecognition_Google:
+                confidence = result['alternative'][0]['confidence']
+                result = result['alternative'][0]['transcript']
```

Comment:
@nakane11 did you also changed  https://github.com/jsk-ros-pkg/jsk_3rdparty/blob/a207a7ab6d3115f48c2bd030f74993dd3eb4512b/ros_speech_recognition/src/ros_speech_recognition/recognize_google_cloud.py#L10 ? it seems this function retuns text, not lists.

If `self.recognize(audio)` returns structure/list in google and text in other engine, how about
```
def a():
    return "hello"
def b():
    return ("hello", 0.2)

for f in a, b:
    r = f()
    if isinstance(r, basestring):
        result = r
        confident = 1.0
    else:
        result = r[0]
        confident = r[1]
    print("{} r:{} c:{}".format(f, result, confident))
```
so that if you have another engine that returns confident, you do not have to change both `speech_recognition_node` and `src/ros_speech_recognition/recognize_XXXXX.py`

If this PR works as it is, I am ok with this PR.

Reply:
The structure of result differs depending on engines, so  I think two options exist.

1. Return raw response and get result and confidence from it in speech_recognition_node
(change speech_recognition_node.py)
```
actual_result=recognize_XXXXX(show_all=True)

if engine == XXXXX:
  result=actual_result['hoge']
  confidence=actual_result['fuga'] 
```
2. Overwrite recognize_XXXXX of speech_recognition library and return result and confidence in the same format
(add recognize_XXXXX.py like google_cloud)
```
def recognize_XXXXX(self, ***):
    *********
    return (result, confidence)

for f in a, b:
    r = f()
    if isinstance(r, basestring):
        result = r
        confident = 1.0
    else:
        result = r[0]
        confident = r[1]
```
Do you think the second is better?

Reply:
If you changed both `speech_recognition_node.py` and `recognize_google_cloud.py`, I think second one is better becuase you only change 1 file.

If you only changed  `speech_recognition_node.py` that's strange. 
Did `google_speech_recognition` returns structure and others returns text, before this PR?
If so, how did google speech recognition return the results? It assums the `result` is text value as follows. 
google speech recognition was not working?

https://github.com/jsk-ros-pkg/jsk_3rdparty/blob/83e55bb2d53707fbe3fae9f1c01e934b5b2a93b8/ros_speech_recognition/scripts/speech_recognition_node.py#L287-L296

Or the return value of `google_speech_recognition` has changed recently?

Reply:
This PR is for google engine, not for google_cloud.
`recognition_google()` returns structure if `show_all` is `true` in `speech_recognition_node.py`.
`show_all` was `false` before this PR, so `result` was text and it was working.

Reply:
I see, what is the difference 
Both requires Google credentials?
Does Google speech recognition ( not clouds ) runs without wifi ?

Reply:
google speech recognition does not need credential and free.
it was google speech api and old service.
it is said that the api will be deprecated, but it continues.

google cloud text-to-speech needs credentials and credit card registration.
it is new version, and it has more features.